Add marginal cost lines and transformers

TimFuermann opened the pull request:
Closes #1350

## Changes proposed in this Pull Request
added marginal cost for lines and transformers (despite passive components). This was asked and could be helpful for the implementation of for example network tariffs. There is a minimum example provided in the minimum_example folder.

## Checklist

- [x ] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [ x] Unit tests for new features were added (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x ] I consent to the release of this PR's code under the MIT license.

TimFuermann committed:
added minimal example and changes to components

TimFuermann committed:
Merge branch 'PyPSA:master' into add-marginal_cost-lines

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

fneum commented:
Careful, it's not that easy. The line and transformer flows regularly have negative values if flows are in opposite direction to operation. In this case, the cost contribution to the objective would be negative, which is not correct.

TimFuermann committed:
use absolute flow in objective for marginal line and trafo costs

TimFuermann committed:
Merge branch 'PyPSA:master' into add-marginal_cost-lines

TimFuermann committed:
Merge branch 'add-marginal_cost-lines' of https://github.com/TimFuermann/PyPSA into add-marginal_cost-lines

TimFuermann committed:
Merge branch 'add-marginal_cost-lines' of https://github.com/TimFuermann/PyPSA into add-marginal_cost-lines

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

fneum reviewed (changes requested):
Logic is correct as I can see, but some further steps necessary to get this merged:

- unit tests for this functionality (should be possible to trace by hand; should capture line/transformer with all-positive, all-negative and alternating flows)
- remove `minimal_example` directory (why was it added?)
- clean up docstring of `define_absolute_auxilaries` (seems to be unchecked AI-generated, does not follow standard (e.g. "Applies to Components"), and is too verbose).
- typo: `define_absolute_auxil*i*aries`
- improved documentation in `lines.csv` and `transformers.csv` how marginal cost are handled.
- adjustments to `v1-docs` branch (I can do that once we decide to merge).
- release notes

TimFuermann committed:
added unit tests, improved documentation and added realease notes

TimFuermann committed:
Merge branch 'add-marginal_cost-lines' of https://github.com/TimFuermann/PyPSA into add-marginal_cost-lines

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

TimFuermann commented:
> Logic is correct as I can see, but some further steps necessary to get this merged:
> 
>  [X]   * unit tests for this functionality (should be possible to trace by hand; should capture line/transformer with all-positive, all-negative and alternating flows)
> 
>  [X]   * remove `minimal_example` directory (why was it added?)
> 
>  [X]   * clean up docstring of `define_absolute_auxilaries` (seems to be unchecked AI-generated, does not follow standard (e.g. "Applies to Components"), and is too verbose).
> 
> [X]    * typo: `define_absolute_auxil*i*aries`
> 
> [X]    * improved documentation in `lines.csv` and `transformers.csv` how marginal cost are handled.
> 
>  * adjustments to `v1-docs` branch (I can do that once we decide to merge).
> 
> [X]    * release notes

@fneum I addressed your comments, sorry for not having the documentation and testing ready from start on. I hope the test and documentation are fine now.

fneum reviewed (changes requested):
I'm generally fine with adding this feature. It just needs a little more documentation:

- a new section here which introduces the new terms in the objective function (As well as new auxiliary constraints and variables): https://docs.pypsa.org/latest/user-guide/optimization/objective/

- Could you add a test that the new cost terms show up in `n.statistics.opex()`.

Also mark any comments that you have already addressed as resolved.

fneum commented:
@gugrimm @TimFuermann is there still interest to finish up this PR?

gugrimm commented:
yes, still interested, sorry :|

MaykThewessen commented:
Three notes, in order of value:

**1. Real-world use case (validation).** We use this feature shape today via a workaround in a PyPSA-derived NL transmission model — cross-border AC lines split into bidirectional `Link` pairs with `marginal_cost = hurdle_rate` (TSO border friction). That workaround removes those corridors from the admittance/PTDF matrix and distorts domestic loop flows; we document the limitation in-code. #1352 lets us drop the conversion and keep the lines AC. Strong +1 on landing this.

**2. Confirmed gap on `n.statistics.opex()` (@fneum, 2025-12-17 review).** Reproduced on the PR branch (`add-marginal_cost-lines @ 1030c79`):

```python
import pypsa, pandas as pd
n = pypsa.Network()
n.set_snapshots(pd.date_range("2025-01-01", periods=3, freq="h"))
n.add("Bus", "A"); n.add("Bus", "B")
n.add("Load", "L_A", bus="A", p_set=1.0)
n.add("Load", "L_B", bus="B", p_set=1.0)
n.add("Generator", "Gen_A", bus="A", p_nom=3.0, marginal_cost=10.0)
n.add("Generator", "Gen_B", bus="B", p_nom=3.0, marginal_cost=100.0)
n.add("Line", "A_B", bus0="A", bus1="B", r=1e-4, x=1e-4, s_nom=100.0)
n.lines.at["A_B", "marginal_cost"] = 5.0
n.optimize(solver_name="highs")

print(n.objective)                # 75.0  (60 gen + 15 branch)
print(n.statistics.opex().sum())  # 60.0  ← branch cost missing
```

Suggested test next to `test_lopf_branch_costs.py`:

```python
def test_opex_includes_branch_marginal_cost():
    # ... build network as above ...
    n.optimize(solver_name="highs")
    opex = n.statistics.opex()
    assert ("Line", "-") in opex.index
    assert opex.loc[("Line", "-")] == pytest.approx(15.0, rel=1e-6)
    assert opex.sum() == pytest.approx(n.objective, rel=1e-6)
```

Mirror for `Transformer` with the same numerics.

**3. KKT note for the `objective/` doc section (@Irieo i, ii — radial case).** For a single line, no cycle (so no KVL multiplier), interior generators, linear tariff τ ≥ 0:

Stationarity on `aux` (the |s| auxiliary): `α + β = τ`, with complementarity selecting `α = τ, β = 0` when `s > 0` and `α = 0, β = τ` when `s < 0`. Stationarity on `s` plus the LMP convention `π_b = -λ_b` then gives

```
π_{b1} − π_{b0} = (α − β) + (μ⁺ − μ⁻) = τ · sign(s) + μ_capacity
```

So in the radial case: uncongested ⇒ the nodal price wedge equals the tariff (signed by flow direction); congested ⇒ tariff plus the capacity dual. Quadratic case replaces τ with `τ + 2q|s|`.

**Caveat for meshed AC**: per-line stationarity picks up an additional `ν_cycle · x_line` term from the KVL constraint, so the wedge is no longer just `τ · sign(s) + μ_capacity` — loop flows redistribute. That's exactly the regime @Irieo's `pypsa.examples.ac_dc_meshed()` test (iv) would pin down. Happy to draft that test + the matching doc paragraph as a follow-up if the author would welcome it.

---

Unrelated CI heads-up: the PR's own tests fail under current pandas master because xarray can't index an `ArrowStringArray` bus index in `pypsa/optimization/constraints.py:914`. Works with `pd.set_option("future.infer_string", False)`. Not introduced by this PR, but tests will be red on CI until that compat lands.

FabianHofmann commented:
@TimFuermann are you planning on working on this further? we could take over, just need this info for planning

TimFuermann commented:
> @TimFuermann are you planning on working on this further? we could take over, just need this info for planning

@FabianHofmann, sorry for the delayed reply. Since this was originally just planned as a small addition for the work of @gugrimm, I think you taking it over and adopting it in the way it makes more sense with the overall framework would be great. Thanks a lot for the help.
